### PR TITLE
fix: C# ref write-back sees through interface and abstract dispatch with a sole implementer (#1356)

### LINKED ISSUE #1356: C#: ref write-back cannot see through interface or abstract dispatch
Found while doing the adversarial pass that #1353 item 4 asked for, rather than by a reviewer. The `ref` write-back proof (#1351, #1355) cannot see through interface dispatch, so a real flow is silently dropped.

## Reproduction

```csharp
public interface IHelper
{
    void Fill(string src, ref string dst);
}

public class Helper : IHelper
{
    public void Fill(string src, ref string dst)
    {
        dst = src;
    }
}

public class Leaky
{
    public void Run()
    {
        var token = Environment.GetEnvironmentVariable("K");
        IHelper helper = new Helper();
        var sink = "";
        helper.Fill(token, ref sink);
        Console.WriteLine(sink);   // no ENV -> STDOUT edge
    }
}
```

Verified failing against the current `main`.

## Cause

The same shape as every other defect in this helper: the INVOKED symbol and the symbol that OWNS the body are different objects. Here `GetSymbolInfo` returns `IHelper.Fill`, whose `DeclaringSyntaxReferences` point at the interface member -- which has no body -- so `CallableBody` returns null, `RefParameterIsWritten` answers false, and the write is treated as not happening. An `abstract` method has the identical problem.

`out` is unaffected: it needs no proof, because the language guarantees definite assignment.

## Why this is not a quick fix, and needs a decision rather than a patch

The implementation is not statically known from the call site. `helper` is declared `IHelper`, so Roslyn's type for it IS the interface. Resolving the actual writer means choosing a policy, and the choice trades the two failure modes against each other:

1. **Sole implementer.** If the solution contains exactly one implementing type, analyse its implementation. This has precedent in this repo: the Go frontend already emits an edge through an interface when there is a SOLE implementer (`_join_go_implements`). Conservative, and consistent with what CGR already does elsewhere.
2. **Any implementer writes it.** MAY semantics -- correct for taint in the abstract, but it fabricates an edge when the implementation actually reached does not write. This helper exists specifically because a read-only `ref` must not invent a flow (the first Greptile finding on #1351), so this option argues against the reason the proof was built.
3. **Leave it.** A missed edge, which is the direction everything else in this helper errs toward.

I lean towards (1), because it matches existing repo behaviour for interface dispatch and keeps the no-fabrication property, but it is a semantics decision rather than a bug fix and it also needs `SymbolFinder`/implementation lookup threaded into `FactCollector`, which currently holds only per-document semantic models plus the tree-owner index added in #1355.

Related: #1187 lists `SymbolFinder` completeness as remaining scope, so this may want to land with that work rather than on its own.

## Resolution (decided in PR #1414)

Option 1, generalized to **must-agreement** so it stays well defined once default interface bodies enter the picture: the write is proven exactly when EVERY body the call can dispatch to writes the parameter. With a sole implementer this is precisely option 1; with several candidates an edge appears only when all of them write, so whichever implementation actually runs performs the write and no flow is ever fabricated (the property option 2 would have broken). Candidates are the declared implementations in loaded source (derived-interface explicit defaults included), plus the invoked member's own default body while any implementer inherits it, and resolution is limited to contracts declared in loaded source, where the implementation set is knowable.

### PR DESCRIPTION
**Pull request by vitali87**
## Summary

- The `ref` write-back proof resolved a dispatched call to the interface or abstract member, whose declaration has no body, so `CallableBody` returned null and a real ENV-to-STDOUT flow through `IHelper.Fill(token, ref sink)` was silently dropped (#1356's verified repro).
- Final policy, a strict generalization of the issue's option 1 that keeps its no-fabrication property: the write is proven exactly when EVERY body the call can dispatch to writes the parameter. With a sole implementer this reduces to option 1; with several candidates the edge exists only when they all write, so whichever implementation actually runs performs the write and nothing is ever invented. The generalization became necessary during review: default interface bodies mean a dispatched member can have several reachable bodies (the live default plus overrides), where "exactly one" is not a well-defined filter.
- Candidate bodies are the declared implementations across the loaded compilations (interfaces themselves included, so a derived interface's explicit default participates), plus the invoked member's own default body while any implementer still inherits it. Resolution happens only for contracts declared in loaded source: references are acyclic, so nothing prebuilt can implement a source contract, while a metadata contract can have invisible implementers and stays unproven. Implementations are deduplicated by source location so multi-targeted compilations cannot distort the candidate set.

## Type of Change

- [x] Bug fix

## Related Issues

Fixes #1356. Related to #1187 (SymbolFinder completeness): this uses the compilations already held by the #1355 tree-owner index instead of threading `SymbolFinder` in.

## Test Plan

- [x] Unit tests pass (`make test-parallel` or `uv run pytest -n auto -m "not integration"`)
- [x] New tests added

Red first, then green with the real Roslyn tool: write-back through a sole interface implementer and a sole abstract override; no flow when the sole implementer only reads; no flow with two implementers when one reads; a default interface method never resolves to an override while an implementer inherits the default; a dead default (every implementer overrides) does not mask the sole override; derived-interface explicit defaults participate in both directions; a metadata-declared contract never resolves to the visible source subset, with a positive direct-call control proving the frontend ran. Full C# frontend surface passes (30 flow tests plus semantic facts and wiring).

## Checklist

- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format
- [x] All pre-commit checks pass (`make pre-commit`)
- [x] No hardcoded strings in non-config/non-constants files
- [x] No `# type: ignore`, `cast()`, `Any`, or `object` type hints
- [x] No new comments or docstrings (code should be self-documenting)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved accuracy when tracking `ref` parameter write-backs through interface and virtual dispatch.
  * Prevented read-only or ambiguous implementations from incorrectly reporting data flows.
  * Correctly handles default interface methods, inherited defaults, derived-interface overrides, metadata-defined interfaces, and generic interface constructions.
  * Ensured results remain reliable when multiple possible implementations are available.

* **Tests**
  * Added comprehensive coverage for dispatch scenarios, abstract and virtual implementations, metadata assemblies, generic calls, inherited defaults, and conditional code paths.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 1]
```diff
@@ -513,6 +525,134 @@ private bool RefParameterIsWritten(
...
+                    .FirstOrDefault(m => SymbolEqualityComparer.Default.Equals(
+                        m.OriginalDefinition, method));
+                // Only where the implementation is DECLARED: a derived type
+                // inherits its base's implementation, and counting it again
+                // there would make a genuinely sole writer look ambiguous.
+                if (member is not null
+                    && type.FindImplementationForInterfaceMember(member)
+                        is IMethodSymbol implementation
+                    && SymbolEqualityComparer.Default.Equals(
+                        implementation.ContainingType, type))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit: SoleDispatchTarget now returns null for any non-abstract member, so a default interface method is never resolved to an override; the call falls through to the pre-existing direct analysis of the default body. Covered red-first by test_a_default_interface_method_never_resolves_to_an_override (Writer overrides and writes, Reader inherits the read-only default, no edge).

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 2]
```diff
@@ -513,6 +525,134 @@ private bool RefParameterIsWritten(
...
+            var seen = new HashSet<(string, int)>();
+            foreach (var type in AllSourceTypes(caller))
+            {
+                var implementation = viaInterface
+                    ? InterfaceImplementation(type, method)
+                    : OverrideOf(type, method);
+                if (implementation is null || implementation.IsAbstract)
+                {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:725]: Applied in the follow-up commit exactly as proposed: the guard is now abstract-members-only, with the viaInterface distinction computed after it. Red-first test added for the inherited-default case.

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 3]
```diff
@@ -513,6 +525,134 @@ private bool RefParameterIsWritten(
...
+                if (implementation.DeclaringSyntaxReferences.FirstOrDefault()
+                    is not { } declared)
+                {
+                    continue;
+                }
+                if (!seen.Add((declared.SyntaxTree.FilePath, declared.Span.Start)))
+                {
+                    continue;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: I believe this one cannot happen: an implementation is only counted when it matches through symbol identity against the caller compilation's member (AllInterfaces/OverriddenMethod compared with SymbolEqualityComparer), and a sibling target framework's compilation binds distinct symbols for the same declaration, so its types never match and never reach the location dedup. Each target's pass therefore resolves within its own compilation and emits or omits its own fact, which then unions across targets exactly as the existing direct-call analysis of an #if-conditional callee already does. The dedup set only guards the cross-project compilation-reference case where symbol identity IS shared and the same declaration is enumerated once per referencing compilation.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 4]
```diff
@@ -513,6 +525,140 @@ private bool RefParameterIsWritten(
...
+            // Abstract members only: a DEFAULT interface method owns a body
+            // and can be the member a call actually lands on (any implementer
+            // that does not override it inherits it), so resolving a sole
+            // override there would attribute the write to a body the call
+            // never reaches. Classic interface members are abstract, so they
+            // still resolve.
+            if (!method.IsAbstract)
+            {
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit by generalizing to a candidate-body model: the dispatch candidates are every declared implementation, plus the default body itself only while some implementer actually inherits it (or nothing overrides it). The write is proven when every candidate writes. Your scenario, a read-only default whose sole implementation writes and which nothing inherits, now resolves to that implementation and emits the flow, covered red-first by test_a_dead_default_body_does_not_hide_the_sole_override; the inherited-default case keeps its no-fabrication behavior.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 5]
```diff
@@ -513,6 +535,162 @@ private bool RefParameterIsWritten(
...
+            {
+                var compilations = new HashSet<Compilation>(_treeOwners.Values) { caller };
+                var types = new List<INamedTypeSymbol>();
+                foreach (var compilation in compilations)
+                {
+                    CollectNamedTypes(compilation.Assembly.GlobalNamespace, types);
+                }
+                _sourceTypes = types;
+            }
+            return _sourceTypes;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:558]: Fixed in the follow-up commit: DispatchCandidates now returns null unless the contract's containing type is declared in loaded source, so a metadata interface or abstract class never resolves to the visible source subset and stays unproven. References are acyclic, so nothing prebuilt can implement a source-declared contract, which keeps the source implementation set complete exactly where resolution still happens. Covered red-first by test_a_metadata_interface_never_resolves_to_source_implementers, which prebuilds the interface into a DLL with dotnet build and references it by HintPath.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 6]
```diff
@@ -513,6 +535,173 @@ private bool RefParameterIsWritten(
...
+                        continue;
+                    }
+                    // Only where the implementation is DECLARED: a derived
+                    // type inherits its base's implementation, which is
+                    // already counted at the base.
+                    if (!SymbolEqualityComparer.Default.Equals(
+                            implementation.ContainingType, type))
+                    {
+                        continue;
+                    }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: I could not reproduce this: AllSourceTypes deliberately includes interfaces themselves, so a derived interface is scanned as a type and its explicit default passes the declared-here filter (implementation.ContainingType == the derived interface == the scanned type), becoming a candidate body. Both directions are now pinned by red-run-verified regression tests on the exact shapes described: test_a_reading_derived_interface_default_blocks_the_base_default_write (writing base default + read-only derived default, no edge) and test_a_writing_derived_interface_default_proves_the_base_member_write (abstract base + writing derived default, edge present). Both pass against this head.

File: codebase_rag/tests/test_csharp_semantic_flow.py [hunk 7]
```diff
@@ -621,3 +621,281 @@ def test_a_read_only_ref_local_function_never_gains_a_flow(tmp_path: Path) -> No
...
+            capture=_CAPTURE_IO,
+        ).run()
+    finally:
+        settings.CSHARP_FRONTEND = previous
+    flows = {
+        (c.args[0][2], c.args[2][2])
+        for c in mock.ensure_relationship_batch.call_args_list
+        if str(c.args[1]) == FLOWS_TO
+    }
+    assert (_ENV_K, _STDOUT) not in flows
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_csharp_semantic_flow.py:906]: Added in the follow-up commit: the fixture now also makes a DIRECT call whose write-back must produce an ENV::K2 edge, asserted before the absence check, so a broken frontend build fails the positive control instead of passing the negative vacuously. The manual settings save/restore also moved to the monkeypatch fixture.

**Comment by vitali87** [comment]
@coderabbitai full review

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@coderabbitai full review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 8]
```diff
@@ -513,6 +535,173 @@ private bool RefParameterIsWritten(
...
+                {
+                    continue;
+                }
+                if (implementation.DeclaringSyntaxReferences.FirstOrDefault()
+                    is not { } declared)
+                {
+                    continue;
+                }
+                if (!seen.Add((declared.SyntaxTree.FilePath, declared.Span.Start)))
+                {
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit: the location dedup is removed entirely, so every per-compilation candidate reaches the must-agreement conjunction and a writing variant can no longer stand in for a read-only one; re-proving an identical body twice is merely redundant. Reproduced red-first with the realizable shape of this finding: a file linked into two projects with different DefineConstants implementing a shared project-referenced interface (test_conditional_variants_of_a_shared_file_are_not_collapsed). A plural-TargetFrameworks fixture would need ref packs for several TFMs; the linked-file layout exercises the identical collapse.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@coderabbitai full review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 9]
```diff
@@ -486,6 +486,28 @@ private bool RefParameterIsWritten(
             var method = (reduced ?? invoked).OriginalDefinition;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:486]: Fixed in the follow-up commit: dispatch now works on the CONSTRUCTED symbol end to end. InterfaceImplementation matches the type's AllInterfaces instance against the invoked member's constructed containing interface and selects the member by constructed identity, so a type implementing IGen<int> and IGen<string> with different bodies answers only for the construction the call went through. Red-first tests cover both directions (read-only construction stays unproven while the writing construction proves).

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 10]
```diff
@@ -513,6 +535,171 @@ private bool RefParameterIsWritten(
...
+            // implementation set: references are acyclic, so nothing prebuilt
+            // can implement a source interface, while a METADATA contract can
+            // have implementers in other referenced assemblies that a source
+            // scan cannot see, and proving a write from the visible subset
+            // could fabricate a flow for a runtime receiver from metadata.
+            if (!method.ContainingType.OriginalDefinition.Locations
+                    .Any(location => location.IsInSource))
+            {
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit, and thanks for pushing back: my acyclicity argument missed assembly unification, where a plugin compiled against an identical-identity assembly genuinely implements the source contract's symbols. DispatchCandidates now also scans every referenced metadata assembly's types (memoized once per run); a metadata implementation joins the candidate list, owns no analyzable body, and therefore fails the proof, keeping the write unproven. Reproduced red-first with a two-stage prebuilt fixture (contract DLL, then a plugin DLL referencing it, unified with the source project via AssemblyName) plus a positive direct-call control.

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 11]
```diff
@@ -513,6 +535,171 @@ private bool RefParameterIsWritten(
             return flow.WrittenInside.Contains(parameter, SymbolEqualityComparer.Default);
         }
 
+        private List<INamedTypeSymbol>? _sourceTypes;
+
+        // Every named type declared in the loaded compilations, memoized: the
+        // dispatch resolution below scans them all, and the set never changes
+        // within a run.
+        private IReadOnlyList<INamedTypeSymbol> AllSourceTypes(Compilation caller)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:546]: Same fix as the sibling thread: referenced metadata assemblies are scanned into the candidate set and their bodiless implementations poison the conjunction, so a source writer can no longer vouch for a metadata receiver. Covered by test_a_prebuilt_plugin_implementation_keeps_the_write_unproven.

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 12]
```diff
@@ -513,6 +535,171 @@ private bool RefParameterIsWritten(
...
+        // (or nothing overrides it), because a default every implementer
+        // overrides can never execute and must not mask the overrides.
+        private IReadOnlyList<IMethodSymbol>? DispatchCandidates(
+            Compilation caller, IMethodSymbol method)
+        {
+            var viaInterface = method.ContainingType.TypeKind == TypeKind.Interface;
+            if (!viaInterface && !method.IsAbstract)
+            {
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:660]: Fixed in the follow-up commit: virtual and overridable-override members now enter dispatch resolution, with the member's own body joining its overrides in the must-agreement conjunction (over-approximating for an abstract containing type, which errs toward a missed edge, never a fabricated one). Red-first tests: a read-only override blocks the virtual base's write, and the edge survives when every override also writes.

File: codebase_rag/tests/test_csharp_semantic_flow.py [hunk 13]
```diff
@@ -621,3 +621,475 @@ def test_a_read_only_ref_local_function_never_gains_a_flow(tmp_path: Path) -> No
...
+        parsers=parsers,
+        queries=queries,
+        capture=_CAPTURE_IO,
+    ).run()
+    flows = {
+        (c.args[0][2], c.args[2][2])
+        for c in mock.ensure_relationship_batch.call_args_list
+        if str(c.args[1]) == FLOWS_TO
+    }
+    assert (_ENV_K, _STDOUT) not in flows
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_csharp_semantic_flow.py:1294]: Added in the follow-up commit: ImplB's Leaky now also calls a non-shared LocalWriter directly, and the test asserts that ENV::K2 edge before the absence check, so the negative cannot pass on a partial load or a broken frontend build.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@coderabbitai full review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 14]
```diff
@@ -513,6 +538,218 @@ private bool RefParameterIsWritten(
...
+                var seenAssemblies = new HashSet<string>(StringComparer.Ordinal);
+                var types = new List<INamedTypeSymbol>();
+                foreach (var compilation in compilations)
+                {
+                    foreach (var assembly
+                        in compilation.SourceModule.ReferencedAssemblySymbols)
+                    {
+                        if (assembly.Locations.Any(location => location.IsInSource))
+                        {
+                            continue;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Partially fixed, with an honest boundary: the follow-up commit walks the FULL reference closure (BFS over each assembly's module references) instead of only direct ReferencedAssemblySymbols, so any transitively referenced assembly the compilation actually RESOLVED now contributes candidates. The remaining gap is not reachable statically: I built your exact shape (app -> Shim.dll -> Plugin.dll with the read-only Reader) and the app compilation never resolves Plugin.dll at all, because csc receives only direct references; the closure then contains a missing-assembly symbol with no types, so no symbol walk can surface Reader. An unresolved assembly's implementations are invisible to every compilation in the workspace, the same way a receiver loaded via Assembly.LoadFrom would be, so the proof's contract remains the compilation-visible implementation set, now covering direct and resolved-transitive metadata implementers.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 15]
```diff
@@ -513,6 +538,234 @@ private bool RefParameterIsWritten(
...
+                        pending.Enqueue(assembly);
+                    }
+                }
+                while (pending.Count > 0)
+                {
+                    var assembly = pending.Dequeue();
+                    if (!seenAssemblies.Add(assembly.Identity.GetDisplayName()))
+                    {
+                        continue;
+                    }
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit: an unresolved (missing-metadata, non-source) assembly symbol now skips before touching seenAssemblies, so a resolved symbol for the same identity seen through another project is always walked regardless of iteration order. A deterministic red test is not constructible for this one because the trigger depends on unspecified HashSet iteration order across compilations; the fix removes the order sensitivity entirely, and the existing resolved-plugin and metadata-contract tests keep covering the collection path.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 16]
```diff
@@ -513,6 +538,243 @@ private bool RefParameterIsWritten(
...
+                    // types and no module references; it must not consume the
+                    // identity, or a RESOLVED symbol for the same assembly
+                    // seen through another project would be skipped.
+                    if (!inSource && assembly.GetMetadata() is null)
+                    {
+                        continue;
+                    }
+                    if (!seenAssemblies.Add(assembly.Identity.GetDisplayName()))
+                    {
+                        continue;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Addressed in the follow-up commit: the visited key is now origin plus identity, so a source assembly and a same-identity resolved metadata assembly are walked independently. Worth noting the concrete risk was theoretical: no single compilation can reference two same-identity assemblies, so a metadata twin's types bind their own copy of the contract and can never be symbol-equal to the source contract used for candidate matching; the key split removes the concern structurally either way.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 17]
```diff
@@ -513,6 +538,249 @@ private bool RefParameterIsWritten(
...
+            // Matched by the CONSTRUCTED interface, never the open definition:
+            // a type implementing IGen<int> and IGen<string> with different
+            // bodies answers only for the construction the call went through.
+            var target = invoked.ConstructedFrom;
+            foreach (var iface in type.AllInterfaces)
+            {
+                if (!SymbolEqualityComparer.Default.Equals(
+                        iface, invoked.ContainingType))
+                {
+                    continue;
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit: an interface instance that still carries type parameters (an open implementer such as Reader<T> : IHelper<T>) now matches by original definition, so its body joins the candidates for every construction; closed implementations keep the exact constructed matching from the earlier fix. Red-first tests cover both directions: the open read-only implementer blocks the writing default, and an open writer proves the constructed call that was previously missed.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 18]
```diff
@@ -513,6 +538,281 @@ private bool RefParameterIsWritten(
...
+            // over-approximating an incompatible construction only adds a
+            // body to the conjunction, erring toward a missed edge.
+            var target = invoked.ConstructedFrom;
+            foreach (var iface in type.AllInterfaces)
+            {
+                var open = ContainsTypeParameters(iface);
+                var matches = open
+                    ? SymbolEqualityComparer.Default.Equals(
+                        iface.OriginalDefinition,
+                        invoked.ContainingType.OriginalDefinition)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:765]: Fixed in the follow-up commit: InterfaceImplementations now yields the implementation from EVERY matching interface instance instead of returning on the first, so an open implementer that also carries an explicit implementation for one construction contributes all reachable bodies to the must-agreement conjunction. Red-first test: Both<T> : IHelper<T>, IHelper<int> where the general body writes and the explicit IHelper<int> body reads; the IHelper<int> call no longer gains an edge.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 19]
```diff
@@ -513,6 +538,285 @@ private bool RefParameterIsWritten(
...
+            INamedTypeSymbol type, IMethodSymbol method)
+        {
+            foreach (var candidate in type.GetMembers(method.Name).OfType<IMethodSymbol>())
+            {
+                for (var overridden = candidate.OverriddenMethod;
+                     overridden is not null;
+                     overridden = overridden.OverriddenMethod)
+                {
+                    if (SymbolEqualityComparer.Default.Equals(
+                            overridden.OriginalDefinition, method))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs]: Fixed in the follow-up commit: OverrideOf now applies the same construction discipline as the interface path, comparing the override chain against the CONSTRUCTED invoked member (with definition matching only for chain entries that still carry type parameters, i.e. open derived classes). Red-first test: IntReader overriding Base<int> no longer blocks the write proven through Base<string>.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs [hunk 20]
```diff
@@ -513,6 +538,296 @@ private bool RefParameterIsWritten(
...
+                if (!matches)
+                {
+                    continue;
+                }
+                var member = iface.GetMembers(invoked.Name)
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(m => open
+                        ? SymbolEqualityComparer.Default.Equals(
+                            m.OriginalDefinition, target.OriginalDefinition)
+                        : SymbolEqualityComparer.Default.Equals(m, target));
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/csharp_frontend/roslyn/Frontend.cs:777]: Fixed in the follow-up commit: open-ness now widens the match from either side. When the invoked interface construction itself carries type parameters (a call inside generic code), every implementation, closed constructions included, joins the must-agreement conjunction, because the call can resolve to any construction at runtime. Red-first test: inside Pump<T>, the read-only explicit IHelper<int> body now blocks the writing open implementation. Closed-vs-closed matching stays exact.

**Comment by vitali87** [comment]
@greptile review
